log/truncating_file_writer.go

```diff
@@ -38,15 +41,36 @@ func NewTruncatingFileWriter(f *os.File, intervals ...time.Duration) TruncatingF
 }
 
 func (w *truncatingFileWriter) Write(p []byte) (n int, err error) {
-	if w.interval.Nanoseconds() > 0 && (time.Now().UnixNano()-w.lastTruncated.UnixNano() >= w.interval.Nanoseconds()) {
-		if err := w.Truncate(); err != nil {
-			w.lastTruncated = time.Now()
-		}
+	now := time.Now()
+	lastTruncated := w.getLastTruncatedUnixNano()
+	if w.interval.Nanoseconds() > 0 && (shouldTruncate(now, lastTruncated, w.interval)) {
+		_ = w.autoTruncate(now) // TODO log the failure to truncate?
 	}
 
 	return w.f.Write(p)
 }
 
+func (w *truncatingFileWriter) getLastTruncatedUnixNano() time.Time {
+	w.mu.RLock()
```

Comment:
no need to lock (pessimistic sync) here. a sync-free optimistic approach at this point would be much faster and will not affect practical correctness, as the `autoTruncate` method is already pessimistic.

Reply:
also the `RWMutex` can be reduced to a regular `Mutex`

Reply:
replacing with `atomic` read/writes for read lock - to protect against race detector alerts, although i agree that a rare random result in the first check does not justify the overhead of a lock